src/worktree.c

```diff
@@ -365,12 +365,25 @@ int git_worktree_add(git_worktree **out, git_repository *repo,
 		if ((err = git_reference_dup(&ref, wtopts.ref)) < 0)
 			goto out;
 	} else {
-		if ((err = git_repository_head(&head, repo)) < 0)
-			goto out;
-		if ((err = git_commit_lookup(&commit, repo, &head->target.oid)) < 0)
-			goto out;
-		if ((err = git_branch_create(&ref, repo, name, commit, false)) < 0)
+		git_buf ref_buf = GIT_BUF_INIT;
```

Comment:
This must be `git_buf_dispose`d of — CI reports a leak.

Reply:
> Hmm, isn't this supposed to be passed as `git_worktree_add_options`'s [ref field](https://libgit2.org/libgit2/#HEAD/type/git_worktree_add_options) ? As far as I can see this changes the behavior, as the old code would have _never_ reused a reference, instead returning an error. Now you can _never_ now if this actually happened, and you can end-up reusing by accident another unrelated reference.

`ref` field looks a bit different to me, it's when you want to checkout `ref` in the worktree.
libgit2 will checkout this `ref` only if it already exists and is not already checked out.

When user doesn't pass `ref` field, current version of `git_worktree_add` creates a new branch with the `name` of the worktree in the main working tree and checkout HEAD in the new worktree.

The only thing this commit is changing is when `ref` field is not passed as an option and branch `refs/heads/<name>` already exists in the main working tree.
It would have error out previously. But with this commit, it will now checkout the commit pointed by `refs/heads/<name>` in the new worktree without any error.

It's the same behavior when you do the following:
`git worktree add ../libgit2-wt`
Git create the new worktree in ../libgit2-wt and create a branch `libgit2-wt` in the main repositoriy.
`rm -Rf ../libgit2-wt`
`git worktree prune`
Here we deleted the `libgit2-wt` and pruned it but Git kept the `libgit2-wt` branch in the main repository.

`git worktree add ../libgit2-wt`
Adding the worktree again works whereas `libgit2-wt` already exists.

This currently doesn't work with libgit2 and that's what I'm trying to fix.

Please let me know if I'm missing something.
Thanks for the review.

Reply:
I fixed that, thanks.